### ISSUE TITLE
Specify necessary types of built-in variables

### ISSUE THREAD

**Issue by sheaf** [KhronosGroup/Vulkan-Docs#953]
Writing a compiler targeting SPIR-V, I was following the Vulkan documentation for built-in variables, and ended up declaring `gl_PerVertex` in a geometry shader as a run-time array of structs, decorating the first member of this struct as the `Position` builtin. This lead to the following error:

> (null)(ERROR / SPEC): msgNum: 3 - geometry shader consumes input location 4294967295.0 which is not written by vertex shader

I'm still not sure why that error pops up, but I eventually figured out that changing the type of the `gl_PerVertex` declaration to an array of definite length (instead of a run-time array) fixes the issue.

I would like the Vulkan specification to be clear about exactly how these built-in variables are expected to appear in SPIR-V. For instance, I noticed that I can leave out fields of the struct without issue (as long as they are not being used, of course). So I would like to know which ways to decorate built-in variables are compatible with SPIR-V, i.e. which specific types will be accepted.

**Comment by johnkslang** [comment]
Vulkan does not allow gl_PerVertex to be a run-time sized array.  Core Vulkan states, in Appendix A: "Vulkan Environment for SPIR-V", under `OpTypeRuntimeArray`:

> `OpTypeRuntimeArray` must only be used for:
>
> * the last member of an `OpTypeStruct` that is in the `StorageBuffer` storage class decorated as Block, or that is in the `Uniform` storage class decorated as `BufferBlock`.

GLSL has the unsized `[ ]` syntax for this, which only temporarily leaves the array unsized, but it is still sized by the time the stage is fully linked, based on other information in the shader.

The GLSL specification, under "Geometry Shader Inputs" has a table that shows the size of the `gl_In[]` array for the different kinds `layout(...)` declarations; e.g., `triangles` will make it size 3.

Basically, nothing should be saying this array can be run-time sized, while GLSL uses empty-size syntax to allow to be link-time sized.

**Comment by sheaf** [comment]
Thanks for the clarification about runtime arrays, which obviously makes complete sense from memory layout considerations.

What I would like is for either the SPIR-V documentation or the Vulkan documentation to explain what the decorations are expected to apply to. Quoting from the SPIR-V validation rules pertaining to builtin decorations:

> A **BuiltIn** Decoration can only be applied as follows:
> * When applied to a structure-type member, all members of that structure type must also be decorated with BuiltIn.
> (No allowed mixing of built-in variables and non-built-in variables within a single structure.)
> * When applied to a structure-type member, that structure type cannot be contained as a member of another structure type.
> * There is at most one object per Storage Class that can contain a structure type containing members decorated with **BuiltIn**, consumed per entry-point.

There is no further information in the section on builtins, e.g. for position:

> **Position**
Output vertex position from a vertex processing Execution Model. See the client API specification for more detail.

In the section on builtins in the Vulkan spec, there are some precisions about Vulkan expects the decorations are applied to, such as:

> Any variable decorated with **Position** must be declared as a four-component vector of 32-bit floating-point values.

However I couldn't find anything in that section that mentions that arrays are sometimes expected. In Appendix A, the only mentions of builtins are:

> Decorations
> * Any **BuiltIn** decoration not listed in Built-In Variables must not be used.
> * Any **BuiltIn** decoration that corresponds only to Vulkan features or extensions that have not been enabled must not be used.

At this point, nothing has mentioned that e.g. the position input for a geometry shader is supposed to be an array, but maybe I've missed something.    

What I would like to see (or be pointed to if I've missed it) is a specification of what the decorations are supposed to apply to, e.g. saying that the `Position` builtin should apply to a member of a structure contained in an array of 4-dimensional 32-bit float vectors of the expected size for the input of a geometry shader, and e.g. for a vertex shader should apply to either a 4-dimensional 32-bit float vector or a struct member of such type which is not contained in an array.    
I think most silly cases are ruled out by the aforementioned validation rules, e.g. a structure containing a structure whose member is decorated as `Position` is ruled out by _"When applied to a structure-type member, that structure type cannot be contained as a member of another structure type."_ But nothing specifically seems to rule out something silly like an array of size 1 containing an array of the expected size containing a struct with a member decorated as `Position` of 4-dimensional 32-bit float vector type. Again, not that one is likely to output such code, but I'm pointing out that I can't find anywhere in the spec that nails down precisely what is expected.

Thanks for your time and assistance.

**Comment by sheaf** [comment]
Ah, I see now that this information is provided in section 14.1 _Shader Input and Output Interfaces_.

> **14.1.3. Interface Matching**
> Tessellation control and mesh shader per-vertex output variables and blocks, and tessellation control, tessellation evaluation, and geometry shader per-vertex input variables and blocks are required to be declared as arrays, with each element representing input or output values for a single vertex of a multi-vertex primitive. For the purposes of interface matching, the outermost array dimension of such variables and blocks is ignored.

This does confirm what I noticed, that the actual sizes of the arrays are ignored as far as this is concerned.

I apologise for missing this!

**Comment by johnkslang** [comment]
No problem; the specs are big and sometimes it is hard to find the key section.

Thanks.
